File: python/Cucumber/CucumberExpressions/argument.py
from __future__ import annotations

from typing import Optional

from Cucumber.CucumberExpressions.group import Group
from Cucumber.CucumberExpressions.parameter_type import ParameterType
from Cucumber.CucumberExpressions.tree_regexp import TreeRegexp
from Cucumber.CucumberExpressions.exceptions.errors import CucumberExpressionError


class Argument:
    def __init__(self, group, parameter_type):
        self._group: Group = group
        self.parameter_type: ParameterType = parameter_type
        self._value = None

    @staticmethod
    def build(
        tree_regexp: TreeRegexp, text: str, parameter_types: list
    ) -> Optional[list[Argument]]:
        group = tree_regexp.match(text)
        if not group:
            return None

        arg_groups = group.children

        if len(arg_groups) != len(parameter_types):
            raise CucumberExpressionError(
                f"Group has {len(arg_groups)} capture groups, but there were {len(parameter_types)} parameter types"
            )

        return [
            Argument(arg_group, parameter_type)
            for parameter_type, arg_group in zip(parameter_types, arg_groups)
        ]

    @property
    def value(self):
        group_values = self.group.values
        return self.parameter_type.transform(group_values)

    @property
    def group(self):
        return self._group

File: elixir/lib/varar/cucumber_expressions/argument.ex
defmodule Varar.CucumberExpressions.Argument do
  @moduledoc """
  A matched argument: a capture group paired with the parameter type that
  transforms it. `value/1` produces the transformed value.
  """

  alias Varar.CucumberExpressions.{Error, Group, ParameterType, TreeRegexp}

  @enforce_keys [:group, :parameter_type]
  defstruct [:group, :parameter_type]

  @type t :: %__MODULE__{group: Group.t(), parameter_type: ParameterType.t()}

  @doc """
  Matches `text` against `tree_regexp` and pairs each top-level capture group
  with its parameter type. Returns `nil` if the text does not match. Raises
  `Varar.CucumberExpressions.Error` if the number of capture groups does not
  match the number of parameter types.
  """
  @spec build(TreeRegexp.t(), String.t(), [ParameterType.t()]) :: [t()] | nil
  def build(tree_regexp, text, parameter_types) do
    case TreeRegexp.match(tree_regexp, text) do
      nil ->
        nil

      group ->
        arg_groups = group.children || []

        if length(arg_groups) != length(parameter_types) do
          raise Error,
            message:
              "Expression /#{Regex.source(tree_regexp.regex)}/ has #{length(arg_groups)} " <>
                "capture groups (#{inspect(Enum.map(arg_groups, & &1.value))}), " <>
                "but there were #{length(parameter_types)} parameter types " <>
                "(#{inspect(Enum.map(parameter_types, & &1.name))})"
        end

        Enum.zip_with(parameter_types, arg_groups, fn parameter_type, arg_group ->
          %__MODULE__{group: arg_group, parameter_type: parameter_type}
        end)
    end
  end

  @doc "The transformed value of this argument."
  def value(%__MODULE__{group: group, parameter_type: parameter_type}) do
    ParameterType.transform(parameter_type, group && Group.values(group))
  end
end
